util/checker/src/cmd/check_stubs_compile.rs

```diff
@@ -0,0 +1,169 @@
+use std::{fs, path::PathBuf, process::Command};
+
+fn make_reserve_copies(modified_exercise: &PathBuf) {
+    let stub_path = modified_exercise.join("src").join("lib.rs");
+
+    let stub_path_copy = modified_exercise.join("src").join("lib.rs.orig");
+
+    let tests_path = modified_exercise.join("tests");
+
+    let tests_path_copy = modified_exercise.join("tests.orig");
+
+    fs::copy(&stub_path, &stub_path_copy)
+        .unwrap_or_else(|_| panic!("Failed to make a reserve copy for {:?}", &stub_path));
+
+    if !tests_path_copy.exists() {
+        fs::create_dir(&tests_path_copy)
+            .unwrap_or_else(|_| panic!("Failed to create {:?} directory", &tests_path_copy));
+    }
+
+    for tests_entry in fs::read_dir(&tests_path)
+        .unwrap_or_else(|_| panic!("Failed to read {:?} directory", &tests_path))
+    {
+        if let Ok(tests_entry) = tests_entry {
+            let entry_path = tests_entry.path();
+            fs::copy(
+                &entry_path,
+                &tests_path_copy.join(entry_path.file_name().unwrap()),
+            ).unwrap_or_else(|_| panic!("Failed to make a reserve copy for {:?}", entry_path));
+        }
+    }
+}
+
+fn remove_copies(modified_exercise: &PathBuf) {
+    let stub_path = modified_exercise.join("src").join("lib.rs");
+
+    let stub_path_copy = modified_exercise.join("src").join("lib.rs.orig");
+
+    let tests_path = modified_exercise.join("tests");
+
+    let tests_path_copy = modified_exercise.join("tests.orig");
+
+    fs::rename(&stub_path_copy, &stub_path).unwrap_or_else(|error| {
+        panic!(
+            "Failed to rename {:?} to {:?}: {}",
+            &stub_path_copy, &stub_path, error
+        )
+    });
+
+    fs::remove_dir_all(&tests_path)
+        .unwrap_or_else(|error| panic!("Failed to delete directory {:?}: {}", &tests_path, error));
+
+    fs::rename(&tests_path_copy, &tests_path).unwrap_or_else(|error| {
+        panic!(
+            "Failed to rename {:?} to {:?}: {}",
+            &tests_path_copy, &tests_path, error
+        )
+    });
+}
+
+fn add_deny_warning_flag(file_path: &PathBuf) {
+    let file_content = fs::read_to_string(&file_path)
+        .unwrap_or_else(|error| panic!("Failed to read file {:?}: {}", &file_path, error));
```

Comment:
It would be nice if this utility used proper error handling from the beginning instead of needing a refactor. The good news is that a custom error enum should be pretty easy for this use case, and you don't actually need the `Failure::Error` trait to use it.

Reply:
You could probably get pretty far, for now, if your custom error type had a single variant which wrapped `io::Result`.

Reply:
Oh, I see down below you've actually done that. Might as well make use of that here, then!

Reply:
Was just working on it. 
I only fear that my approach of using `map_err` to construct custom Error from the `io::Error` introduces too much boilerplate. Maybe there is a better way?

Reply:
Well, that's where the `Error` trait would come in handy. Probably the simplest way to do this without external crates:

- [x] define your custom `Error` enum
- [ ] impl `From<E>` for each error type such as `io::Error` that you wish to wrap
- [ ] handle errors like `.map_err(|e| e.into())?`

The only way to get much shorter is to write a little macro.

Reply:
Sorry, should have specified the problem.

I have made a wrapper around `io::Error` like this: `IOCommandError(String, io::Error)`, where `String` is the failed command. 
So, in order to construct the above wrapper, I have to use code like that:

```rust
map_err(|io_error| {                                                                                                                     
            OSInteractionError::IOCommandError(                                                                                                   
                "git rev-parse --abbrev-ref HEAD".to_string(),                                                                                    
                io_error,                                                                                                                         
            )
```    
which, I fear, is repetitive and quite large when used a lot.

I guess I will just write a macro for it later.

Reply:
Ah. Yeah, in that case, the macro is probably the right way to go. Might as well write it now, it'll prevent you from needing to write all those big `map_err` clauses manually.

Reply:
You should look into the [`$crate`](https://doc.rust-lang.org/1.15.0/book/macros.html#the-variable-crate) pseudo-variable; it's meant to solve that problem. It's also weirdly-poorly documented these days; the language team is trying to get away from it.